File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {

        int left = 0, right = s.length();
        int ans = 0;
        while(left<right){
            int k = left;
            Map<Character,Integer> mp = new HashMap<>();
            while(left<right && !mp.containsKey(s.charAt(left))){
                mp.put(s.charAt(left),left);
                left++;
            }
            ans = Math.max(ans,left-k);
            if(left<right) left = mp.get(s.charAt(left))+1;
                                
        }                
        return ans;

    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        length = len(s)
        left, right = 0, length

        maxCount = 0

        while(left < right):
            ind = {}
            k = left
            while(k < right and s[k] not in ind):
                ind[s[k]] = k
                k += 1
            maxCount = max(maxCount, k - left)
            if k == right:
                return maxCount
            left = ind[s[k]] + 1
        return maxCount